processing/src/main/java/io/druid/query/groupby/GroupByQuery.java

```diff
@@ -302,41 +300,62 @@ public int compare(Row lhs, Row rhs)
             @Override
             public int compare(Row lhs, Row rhs)
             {
-              final int timeCompare;
-
-              if (granular) {
-                timeCompare = Longs.compare(
-                    granularity.truncate(lhs.getTimestampFromEpoch()),
-                    granularity.truncate(rhs.getTimestampFromEpoch())
-                );
-              } else {
-                timeCompare = Longs.compare(
-                    lhs.getTimestampFromEpoch(),
-                    rhs.getTimestampFromEpoch()
-                );
-              }
+              final int timeCompare = timeComparator.compare(lhs, rhs);
 
               if (timeCompare != 0) {
                 return timeCompare;
               }
 
-              for (DimensionSpec dimension : dimensions) {
-                final int dimCompare = naturalNullsFirst.compare(
-                    lhs.getRaw(dimension.getOutputName()),
-                    rhs.getRaw(dimension.getOutputName())
-                );
-                if (dimCompare != 0) {
-                  return dimCompare;
-                }
-              }
-
-              return 0;
+              return dimComparator.compare(lhs, rhs);
             }
           }
       );
     }
   }
 
+  private Comparator<Row> getTimeComparator(boolean granular)
+  {
+    if (granular) {
+      return new Comparator<Row>()
+      {
+        @Override
+        public int compare(Row lhs, Row rhs)
+        {
+          return Longs.compare(
+              granularity.truncate(lhs.getTimestampFromEpoch()),
+              granularity.truncate(rhs.getTimestampFromEpoch())
+          );
+        }
+      };
+    } else {
+      return NON_GRANULAR_TIME_COMP;
+    }
+  }
+
+  private Comparator<Row> getDimComparator()
```

Comment:
Why not just a method for comparing rows in `GroupByQuery` and then calling this method from a comparator, returned from `getRowOrdering()`? Wrapping this functionality into another comparator object seems needless to me.